src/Future.js

```diff
@@ -15,7 +15,35 @@ Future.prototype.map = function(f) {
 
 // apply
 Future.prototype.ap = function(m) {
-    return this.chain(function(f) { return m.map(f); });
+
+  return new Future(function(rej, res) {
+    var err, applyFn, val;
+
+    function resolveIfDone() {
+      if (applyFn != null && val != null) {
+        return res(applyFn(val));
+      }
+    };
+
+    function rejectIfFirst(e) {
+      if (err == null) {
+        err = e;
+        rej(err);
+      }
+    };
+
+    this.fork(rejectIfFirst, function(fn) {
+      applyFn = fn;
+      resolveIfDone();
+    });
+
+    m.fork(rejectIfFirst, function(v) {
+      val = v;
+      resolveIfDone();
+    });
+
+  }.bind(this));
```

Comment:
this means >= IE 9 --not a huge deal, but it is pretty easy to keep a reference to original `this` in closure so `bind` is probably not necessary here.